<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
  <title>Using Nim to count sequence-motifs in a BAM - genomics dev blog</title>
  <link rel="alternate" hreflang="en" href="/" />

<meta name="renderer" content="webkit" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>

<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />

<meta name="theme-color" content="#f8f5ec" />
<meta name="msapplication-navbutton-color" content="#f8f5ec">
<meta name="apple-mobile-web-app-capable" content="yes">
<meta name="apple-mobile-web-app-status-bar-style" content="#f8f5ec">


<meta name="author" content="Brent" />
  <meta name="description" content="This is the 2nd post describing mundane uses of nim in day-to-day genomics.
The first post is here.
For today&amp;rsquo;s mundane task, a colleague asked me to count the occurence of 2 k-mers of length 39 in each of 603 BAMs of 60X coverage. We don&amp;rsquo;t care about partial matches or allowing mismatches so this is a pretty simple task. There are more efficient computational methods for this, but I wrote the simplest version, verified that it ran in &amp;lt; 2 hours for a single bam and then ran it overnight on a single machine with 64 CPUs and sent the results in the morning." />
<meta name="keywords" content="nim" />







<meta name="generator" content="Hugo 0.38-DEV" />


<link rel="canonical" href="/post/nim-mundane2/" />

<link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
<link rel="icon" href="/favicon.ico" />
<link rel="manifest" href="/manifest.json">
<link rel="mask-icon" href="/safari-pinned-tab.svg" color="#5bbad5">




<link href="/dist/jane.min.css?v=2.7.0" rel="stylesheet">
<link href="/lib/fancybox/jquery.fancybox-3.1.20.min.css" rel="stylesheet">




<meta property="og:title" content="Using Nim to count sequence-motifs in a BAM" />
<meta property="og:description" content="This is the 2nd post describing mundane uses of nim in day-to-day genomics.
The first post is here.
For today&rsquo;s mundane task, a colleague asked me to count the occurence of 2 k-mers of length 39 in each of 603 BAMs of 60X coverage. We don&rsquo;t care about partial matches or allowing mismatches so this is a pretty simple task. There are more efficient computational methods for this, but I wrote the simplest version, verified that it ran in &lt; 2 hours for a single bam and then ran it overnight on a single machine with 64 CPUs and sent the results in the morning." />
<meta property="og:type" content="article" />
<meta property="og:url" content="/post/nim-mundane2/" />



<meta property="article:published_time" content="2018-10-04T09:23:24-06:00"/>

<meta property="article:modified_time" content="2018-10-04T09:23:24-06:00"/>











<meta itemprop="name" content="Using Nim to count sequence-motifs in a BAM">
<meta itemprop="description" content="This is the 2nd post describing mundane uses of nim in day-to-day genomics.
The first post is here.
For today&rsquo;s mundane task, a colleague asked me to count the occurence of 2 k-mers of length 39 in each of 603 BAMs of 60X coverage. We don&rsquo;t care about partial matches or allowing mismatches so this is a pretty simple task. There are more efficient computational methods for this, but I wrote the simplest version, verified that it ran in &lt; 2 hours for a single bam and then ran it overnight on a single machine with 64 CPUs and sent the results in the morning.">


<meta itemprop="datePublished" content="2018-10-04T09:23:24-06:00" />
<meta itemprop="dateModified" content="2018-10-04T09:23:24-06:00" />
<meta itemprop="wordCount" content="403">



<meta itemprop="keywords" content="nim-mundane," />
<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="Using Nim to count sequence-motifs in a BAM"/>
<meta name="twitter:description" content="This is the 2nd post describing mundane uses of nim in day-to-day genomics.
The first post is here.
For today&rsquo;s mundane task, a colleague asked me to count the occurence of 2 k-mers of length 39 in each of 603 BAMs of 60X coverage. We don&rsquo;t care about partial matches or allowing mismatches so this is a pretty simple task. There are more efficient computational methods for this, but I wrote the simplest version, verified that it ran in &lt; 2 hours for a single bam and then ran it overnight on a single machine with 64 CPUs and sent the results in the morning."/>

<!--[if lte IE 9]>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/classlist/1.1.20170427/classList.min.js"></script>
<![endif]-->

<!--[if lt IE 9]>
  <script src="https://cdn.jsdelivr.net/npm/html5shiv@3.7.3/dist/html5shiv.min.js"></script>
  <script src="https://cdn.jsdelivr.net/npm/respond.js@1.4.2/dest/respond.min.js"></script>
<![endif]-->


</head>
<body>
  <div id="mobile-navbar" class="mobile-navbar">
  <div class="mobile-header-logo">
    <a href="/" class="logo">bpbio</a>
  </div>
  <div class="mobile-navbar-icon">
    <span></span>
    <span></span>
    <span></span>
  </div>
</div>
<nav id="mobile-menu" class="mobile-menu slideout-menu">
  <ul class="mobile-menu-list">
    <a href="/">
        <li class="mobile-menu-item">Home</li>
      </a><a href="/post/">
        <li class="mobile-menu-item">Archives</li>
      </a><a href="/tags/">
        <li class="mobile-menu-item">Tags</li>
      </a><a href="/categories/">
        <li class="mobile-menu-item">Categories</li>
      </a><a href="/about/">
        <li class="mobile-menu-item">About</li>
      </a>
  </ul>
</nav>

  <header id="header" class="header container">
    <div class="logo-wrapper">
  <a href="/" class="logo">bpbio</a>
</div>

<nav class="site-navbar">
  <ul id="menu" class="menu">
    <li class="menu-item">
        <a class="menu-item-link" href="/">Home</a>
      </li><li class="menu-item">
        <a class="menu-item-link" href="/post/">Archives</a>
      </li><li class="menu-item">
        <a class="menu-item-link" href="/tags/">Tags</a>
      </li><li class="menu-item">
        <a class="menu-item-link" href="/categories/">Categories</a>
      </li><li class="menu-item">
        <a class="menu-item-link" href="/about/">About</a>
      </li>
  </ul>
</nav>
  </header>

  <div id="mobile-panel">
    <main id="main" class="main bg-llight">
      <div class="content-wrapper">
        <div id="content" class="content container">
          <article class="post bg-white">
    
    <header class="post-header">
      <h1 class="post-title">Using Nim to count sequence-motifs in a BAM</h1>

      <div class="post-meta">
        <span class="post-time"> 2018-10-04 </span>
        <div class="post-category">
            
              <a href="/categories/nim/"> nim </a>
            
          </div>
        <span class="more-meta"> 403 words </span>
        <span class="more-meta"> 2 min read </span>
      </div>
    </header>

    
    

    
    <div class="post-content">
      <p>This is the 2nd post describing mundane uses of <a href="https://nim-lang.org">nim</a> in day-to-day genomics.</p>

<p>The first post is <a href="https://brentp.github.io/post/nim-hts-example/">here</a>.</p>

<p>For today&rsquo;s mundane task, a colleague asked me to count the occurence of 2 k-mers of length 39 in
each of 603 BAMs of 60X coverage. We don&rsquo;t care about partial matches or allowing mismatches so this
is a pretty simple task. There are more efficient computational methods for this, but I wrote the
simplest version, verified that it ran in &lt; 2 hours for a single bam and then ran it overnight on
a single machine with 64 CPUs and sent the results in the morning.</p>

<p>The kmers we actually used are redacted to protect the innocent.</p>

<p>The code does the following:</p>

<ul>
<li>implements a reverse-complement function and uses it on the 2 query sequences.</li>
<li>opens a BAM file for the path that is given as an argument to the program.</li>
<li>iterates over the bam and counts the presence of each forward or reverse-complemented kmer.</li>
<li>prints out the bam and the count of times each kmer was observed.</li>
</ul>

<p>Again, there are better algorithms for sequencer- matching (though the string-matching implementation in nim must be pretty efficient),
but this took about 15 minutes to code and was fast enough to set running. Here is the code with explanatory comments:</p>

<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-nim" data-lang="nim"><span style="color:#f92672">import</span> os
<span style="color:#f92672">import</span> strutils
<span style="color:#f92672">import</span> hts
<span style="color:#f92672">import</span> kmer

<span style="color:#66d9ef">var</span> aseq <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;some other dna sequence&#34;</span>.toUpperAscii
<span style="color:#66d9ef">var</span> bseq <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;some other DNA sequence&#34;</span>.toUpperAscii

<span style="color:#75715e"># write a quick reverse-complement function</span>
<span style="color:#66d9ef">proc </span><span style="color:#a6e22e">complement</span>(s:<span style="color:#66d9ef">char</span>): <span style="color:#66d9ef">char</span> {.inline.} <span style="color:#f92672">=</span>
    <span style="color:#66d9ef">if</span> s <span style="color:#f92672">==</span> <span style="color:#e6db74">&#39;C&#39;</span>:
        <span style="color:#66d9ef">return</span> <span style="color:#e6db74">&#39;G&#39;</span>
    <span style="color:#66d9ef">elif</span> s <span style="color:#f92672">==</span> <span style="color:#e6db74">&#39;G&#39;</span>:
        <span style="color:#66d9ef">return</span> <span style="color:#e6db74">&#39;C&#39;</span>
    <span style="color:#66d9ef">elif</span> s <span style="color:#f92672">==</span> <span style="color:#e6db74">&#39;A&#39;</span>:
        <span style="color:#66d9ef">return</span> <span style="color:#e6db74">&#39;T&#39;</span>
    <span style="color:#66d9ef">elif</span> s <span style="color:#f92672">==</span> <span style="color:#e6db74">&#39;T&#39;</span>:
        <span style="color:#66d9ef">return</span> <span style="color:#e6db74">&#39;A&#39;</span>
    <span style="color:#66d9ef">else</span>:
        <span style="color:#66d9ef">return</span> s

<span style="color:#66d9ef">proc </span><span style="color:#a6e22e">reverse_complement</span>(xs: <span style="color:#66d9ef">string</span>): <span style="color:#66d9ef">string</span> <span style="color:#f92672">=</span>
  result <span style="color:#f92672">=</span> newString(xs.len)
  <span style="color:#66d9ef">for</span> i, x <span style="color:#f92672">in</span> xs:
    <span style="color:#75715e"># high == len - 1</span>
    result<span style="color:#f92672">[</span>xs.high<span style="color:#f92672">-</span>i<span style="color:#f92672">]</span> <span style="color:#f92672">=</span> complement(x)

<span style="color:#66d9ef">var</span> raseq <span style="color:#f92672">=</span> aseq.reverse_complement
<span style="color:#66d9ef">var</span> rbseq <span style="color:#f92672">=</span> bseq.reverse_complement

<span style="color:#66d9ef">var</span> path <span style="color:#f92672">=</span> commandLineParams()<span style="color:#f92672">[</span><span style="color:#ae81ff">0</span><span style="color:#f92672">]</span>
<span style="color:#66d9ef">var</span> bam: Bam
open(bam, path, threads<span style="color:#f92672">=</span><span style="color:#ae81ff">2</span>)

<span style="color:#66d9ef">var</span> acount <span style="color:#f92672">=</span> <span style="color:#ae81ff">0</span>
<span style="color:#66d9ef">var</span> bcount <span style="color:#f92672">=</span> <span style="color:#ae81ff">0</span>

<span style="color:#75715e"># this gets filled with the sequence for each read.</span>
<span style="color:#66d9ef">var</span> sequence: <span style="color:#66d9ef">string</span>
<span style="color:#66d9ef">for</span> aln <span style="color:#f92672">in</span> bam:
    <span style="color:#75715e"># check flags to avoid double-counting</span>
    <span style="color:#66d9ef">if</span> aln.flag.dup: <span style="color:#66d9ef">continue</span>
    <span style="color:#66d9ef">if</span> aln.flag.supplementary: <span style="color:#66d9ef">continue</span>
    <span style="color:#75715e"># fill the `sequence` string with the values from this alignment</span>
    <span style="color:#66d9ef">discard</span> aln.sequence(sequence)

    acount <span style="color:#f92672">+=</span> sequence.count(aseq)
    acount <span style="color:#f92672">+=</span> sequence.count(raseq)

    bcount <span style="color:#f92672">+=</span> sequence.count(bseq)
    bcount <span style="color:#f92672">+=</span> sequence.count(rbseq)

<span style="color:#66d9ef">var</span> base <span style="color:#f92672">=</span> path.split(<span style="color:#e6db74">&#39;/&#39;</span>)
<span style="color:#75715e"># just print the final file, not the full directory</span>
echo base<span style="color:#f92672">[</span>base.high<span style="color:#f92672">]</span> <span style="color:#f92672">&amp;</span> <span style="color:#e6db74">&#34;</span><span style="color:#ae81ff">\t</span><span style="color:#e6db74">&#34;</span> <span style="color:#f92672">&amp;</span> <span style="color:#f92672">$</span>acount <span style="color:#f92672">&amp;</span> <span style="color:#e6db74">&#34;</span><span style="color:#ae81ff">\t</span><span style="color:#e6db74">&#34;</span> <span style="color:#f92672">&amp;</span> <span style="color:#f92672">$</span>bcount
<span style="color:#75715e">#</span></code></pre></div>

    </div>

    
    

    
    

    <footer class="post-footer">
      <div class="post-tags">
          
          <a href="/tags/nim-mundane/">nim-mundane</a>
          
        </div>

      
      <nav class="post-nav">
        
        
          <a class="next" href="/post/nim-hts-example/">
            <span class="next-text nav-default">Nim Hts Example</span>
            <span class="prev-text nav-mobile">Next</span>
            <i class="iconfont icon-right"></i>
          </a>
      </nav>
    </footer>
    
  </article>
        </div>
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="social-links">
      <a href="mailto:bpederse@gmail.com" class="iconfont icon-email" title="email"></a>
      <a href="https://twitter.com/brent_p" class="iconfont icon-twitter" title="twitter"></a>
      <a href="https://github.com/brentp" class="iconfont icon-github" title="github"></a>
  <a href="/index.xml" type="application/rss+xml" class="iconfont icon-rss" title="rss"></a>
</div>

<div class="copyright">
  <span class="power-by">
    Powered by <a class="hexo-link" href="https://gohugo.io">Hugo</a>
  </span>
  <span class="division">|</span>
  <span class="theme-info">
    Theme - <a class="theme-link" href="https://github.com/xianmin/hugo-theme-jane">Jane</a>
  </span>

  <span class="copyright-year">
    &copy; 
    2018
    <span class="author">brentp</span>
  </span>
</div>

    </footer>

    <div class="back-to-top" id="back-to-top">
      <i class="iconfont icon-up"></i>
    </div>
  </div>
  
<script src="/lib/highlight/highlight.pack.js?v=20171001"></script>
<script type="text/javascript" src="/lib/jquery/jquery-3.2.1.min.js"></script>
  <script type="text/javascript" src="/lib/slideout/slideout-1.0.1.min.js"></script>
  <script type="text/javascript" src="/lib/fancybox/jquery.fancybox-3.1.20.min.js"></script>
<script type="text/javascript" src="/dist/jane.min.js?v=2.7.0"></script>





</body>
</html>
